### PR TITLE
Checkstyle plugin rules defined in the root project

### PR DESCRIPTION
**Pull request by bmuschko**
### Context

See https://github.com/gradle/gradle/issues/2765 for more context. This pull request replaces https://github.com/gradle/gradle/pull/2811 and add more testing + polishing.

### CODE REVIEW

**Comment by oehme** [comment]
This would be a breaking change. I suggest we use some lazy logic which looks in both locations and add a deprecation warning for the old location.

**Comment by blindpirate** [comment]
+1 with @oehme .

This is a big change, especially for the multi-project with `config.xml` under both root project and child project. This change will result in subproject's config file overwritten silently by root project's config file.

**Comment by bmuschko** [comment]
@oehme @blindpirate Thanks for the feedback. I will make the relevant changes.

**Comment by bmuschko** [comment]
@oehme @blindpirate I made additional changes based on your comments. Could you have another look?

**Comment by blindpirate** [comment]
Thanks @bmuschko for the work. I'm not sure if the deprecation should be done - personally I think that would affect many projects and we can support config files on both locations. Could you elaborate on why you think this should be deprecated?

In addition, the configuration directory detection is not lazy - is there any possibility that a Checkstyle config file is generated by some tasks?

**Comment by bmuschko** [comment]
> I'm not sure if the deprecation should be done - personally I think that would affect many projects and we can support config files on both locations. Could you elaborate on why you think it's necessary to deprecate the old subproject location?

It's unlikely that a large majority of users will read the release notes in detail. Rendering a concrete message will inform and hopefully prompt users to switch to the single root project configuration directory immediately after upgrading. Without the deprecation warning we'll simply break projects with Gradle 5.0 by only reading from the root project directory.

> In addition, the configuration directory detection is not lazy - is there any possibility that a Checkstyle config file is generated by some tasks?

No, the files has to be provided by the user. I am not aware of the plugin generating the file automatically. My assumption is that the file exists before the build is invoked.

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java [hunk 1]
```diff
@@ -52,7 +54,13 @@ protected CodeQualityExtension createExtension() {
         extension = project.getExtensions().create("checkstyle", CheckstyleExtension.class, project);
         extension.setToolVersion(DEFAULT_CHECKSTYLE_VERSION);
 
-        extension.setConfigDir(project.file("config/checkstyle"));
+        if (usesSubprojectCheckstyleConfiguration()) {
+            DeprecationLogger.nagUserOfDeprecated("Setting the Checkstyle configuration file under 'config/checkstyle' of a sub project", "Use the root project's 'config/checkstyle' directory instead");
+            extension.setConfigDir(project.file(CONFIG_DIR_NAME));
```
Comment (oehme) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java]: This could be done using convention mapping instead of a setter call. Then it would be lazy, avoiding IO at configuration time.

File: subprojects/docs/src/docs/release/notes.md [hunk 2]
```diff
@@ -53,6 +53,14 @@ The following are the newly deprecated items in this Gradle release. If you have
 ### Example deprecation
 -->
 
+### Change of default Checkstyle configuration directory
+
+Configuring rules consumed by the Checkstyle plugin in a multi-project build used to be tedious.
+Every sub project needs to provide a configuration file even if the content is exactly the same.
```
Comment (oehme) [inline subprojects/docs/src/docs/release/notes.md]: Not really, the user just had to change the file location in the configuration. I don't think people were actually so helpless that they put the same file in all locations.

File: subprojects/docs/src/docs/release/notes.md [hunk 3]
```diff
@@ -53,6 +53,14 @@ The following are the newly deprecated items in this Gradle release. If you have
...
+With this release of Gradle, the Checkstyle configuration file only needs to be provided in the directory `config/checkstyle` of the root project and automatically applies to all sub projects.
+Setting the Checkstyle configuration file per sub project has been deprecated and is scheduled to be removed with the next major Gradle version.
```
Comment (oehme) [inline subprojects/docs/src/docs/release/notes.md]: This is misleading. We didn't deprecate setting it per subproject. This should still be valid if the user explicitly does so.

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java [hunk 4]
```diff
@@ -52,7 +54,13 @@ protected CodeQualityExtension createExtension() {
         extension = project.getExtensions().create("checkstyle", CheckstyleExtension.class, project);
         extension.setToolVersion(DEFAULT_CHECKSTYLE_VERSION);
 
-        extension.setConfigDir(project.file("config/checkstyle"));
+        if (usesSubprojectCheckstyleConfiguration()) {
+            DeprecationLogger.nagUserOfDeprecated("Setting the Checkstyle configuration file under 'config/checkstyle' of a sub project", "Use the root project's 'config/checkstyle' directory instead");
```
Comment (oehme) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java]: ❌ This warning will be emitted eagerly even if the user really wants a checkstyle file specific to this subproject. We should use a lazy approach instead.

File: subprojects/docs/src/docs/release/notes.md (hunk 2 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/release/notes.md]: That's exactly what the reporter of the issue said if I understand him correctly: https://github.com/gradle/gradle/issues/2765. He also mentions that the default can be changed but I am not sure if that's what he actually did to work around the issue. I can expand on the description here by saying that you can also set the `configDir` property.

Comment (oehme) [reply subprojects/docs/src/docs/release/notes.md]: Quoting from the issue: 

> Currently every time I set up a multi-project build using Gradle I need to have this boilerplate code overriding configFile to rootProject.

File: subprojects/docs/src/docs/release/notes.md (hunk 3 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/release/notes.md]: Removed the sentence. However, the deprecation message indicates that if I'd read the message as a user. Sure, you can directly point to it but nevertheless the message is "don't do that". Let me know if you think that the message should be changed as well.

File: subprojects/docs/src/docs/release/notes.md (hunk 2 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/release/notes.md]: Corrected the wording.

File: subprojects/docs/src/docs/release/notes.md (hunk 3 above)
Comment (oehme) [reply subprojects/docs/src/docs/release/notes.md]: The message shouldn't be shown when the user sets it explicitly.

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java (hunk 1 above)
Comment (bmuschko) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java]: Done. Changed the type in the extension to `Property` and exposed a new method that takes a `Provider` so that the value be be set from the plugin implementation.

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java (hunk 4 above)
Comment (bmuschko) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java]: Can you elaborate how a lazy approach would prevent us from rendering the message? Maybe this speaks for not rendering a deprecation message at all as we want to allow user to set the location explicitly via `configDir` and don't really deprecate the sub project's location. The only thing that changed was the default location.

File: subprojects/docs/src/docs/release/notes.md (hunk 3 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/release/notes.md]: Added a test case that demonstrates the functionality.

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java [hunk 5]
```diff
@@ -37,10 +39,11 @@
     private int maxErrors;
     private int maxWarnings = Integer.MAX_VALUE;
     private boolean showViolations = true;
-    private File configDir;
+    private Property<File> configDir;
 
     public CheckstyleExtension(Project project) {
         this.project = project;
+        configDir = project.getObjects().property(File.class);
```
Comment (oehme) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java]: ❌ Should be a [directory property](https://github.com/gradle/gradle/blob/master/subprojects/core-api/src/main/java/org/gradle/api/file/ProjectLayout.java#L64).

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java [hunk 6]
```diff
@@ -113,7 +116,18 @@ public File getConfigDir() {
...
+    }
+
+    /**
+     * Set the provider for calculating the configuration directory.
+     *
+     * @param configDir Configuration directory provider
+     * @since 4.7
+     */
+    @Incubating
+    public void setConfigDir(Provider<File> configDir) {
```
Comment (oehme) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java]: ❌ Let's use the recommended idiom for providers: A getter returning the `DirectoryProperty`, so users can call get/set on it as needed. At some point we can then deprecate the old getters/setters.

We'll need a different name of course, could be `getConfigurationDirectory`

File: subprojects/docs/src/docs/release/notes.md [hunk 7]
```diff
@@ -53,6 +53,14 @@ The following are the newly deprecated items in this Gradle release. If you have
 ### Example deprecation
 -->
 
+### Change of default Checkstyle configuration directory
+
+Configuring rules consumed by the Checkstyle plugin in a multi-project build used to be tedious.
```
Comment (oehme) [inline subprojects/docs/src/docs/release/notes.md]: ⭕️ I think we can drop the whole initial paragraph.. We're saying everything twice right now and might even confuse users with phrases like "users have to reconfigure".

Saying "it used to be tedious" is a bit over the top in my opinion, it was a one-line change to reconfigure it.

File: subprojects/docs/src/docs/release/notes.md [hunk 8]
```diff
@@ -53,6 +53,14 @@ The following are the newly deprecated items in this Gradle release. If you have
...
+It's a very common scenario that all sub projects of a multi-project build use the same configuration file.
+Users have to reconfigure the default value of the property `configDir` to point to the shared location hosting the configuration file. 
+With this release of Gradle, the Checkstyle configuration file only needs to be provided in the directory `config/checkstyle` of the root project and automatically applies to all sub projects without having to set a new location for the `configDir` property.
```
Comment (oehme) [inline subprojects/docs/src/docs/release/notes.md]: Instead of saying "only needs to be provided" I'd rephrase to "The configuration files are now automatically found in the root project and applied to all subprojects without having to...."

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java (hunk 5 above)
Comment (bmuschko) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java]: I think at the moment there's not a way to set the value of a `DirectoryProperty` lazily e.g. with a `Callable`. The `set` method just takes a `File` instance. In turn that would render the deprecation message again even if we don't need it. Is there a way to defer the evaluation of the value with `DirectoryProperty`?

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java (hunk 6 above)
Comment (bmuschko) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java]: See comment above.

File: subprojects/docs/src/docs/release/notes.md (hunk 7 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/release/notes.md]: done

File: subprojects/docs/src/docs/release/notes.md (hunk 8 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/release/notes.md]: done

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java (hunk 5 above)
Comment (oehme) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java]: It`s a `Property<Directory>`, so you can use a `Provider<Directory>`.

Comment (bmuschko) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java]: Done though I feel like the code looks more complex now.

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java [hunk 9]
```diff
@@ -119,14 +121,25 @@ public void setConfigDir(File configDir) {
...
+
     /**
      * Set the provider for calculating the configuration directory.
      *
      * @param configDir Configuration directory provider
      * @since 4.7
      */
     @Incubating
-    public void setConfigDir(Provider<File> configDir) {
+    public void setConfigurationDirectory(Provider<Directory> configDir) {
```
Comment (oehme) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java]: :x: this should be removed.

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java [hunk 10]
```diff
@@ -113,7 +118,29 @@ public File getConfigDir() {
...
+    }
+
+    /**
+     * Set the provider for calculating the configuration directory.
+     *
+     * @param configDir Configuration directory provider
+     * @since 4.7
+     */
+    @Incubating
+    public void setConfigurationDirectory(Provider<Directory> configDir) {
```
Comment (oehme) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java]: :x: This method shouldn't exist.

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java [hunk 11]
```diff
@@ -52,7 +56,22 @@ protected CodeQualityExtension createExtension() {
         extension = project.getExtensions().create("checkstyle", CheckstyleExtension.class, project);
         extension.setToolVersion(DEFAULT_CHECKSTYLE_VERSION);
 
-        extension.setConfigDir(project.file("config/checkstyle"));
+        extension.setConfigurationDirectory(project.provider(new Callable<Directory>() {
+            @Override
+            public Directory call() {
+                DirectoryProperty configDir = project.getLayout().directoryProperty();
```
Comment (oehme) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java]: :x: You don't want to create a property here, but a simple directory.

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java (hunk 10 above)
Comment (bmuschko) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java]: That makes all sense. However, I am still not clear on how to defer rendering the deprecation message if set through the extension by the user without a `Callable`. Let's just chat tomorrow. That'll get us to the goal faster.

Comment (oehme) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java]: You don't need this method for that, you can just call `getConfigurationDirectory.set()` (which is all this method does anyway.

Comment (bmuschko) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java]: I guess I misunderstand one of your previous comments. done.

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java (hunk 11 above)
Comment (bmuschko) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java]: done

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java [hunk 12]
```diff
@@ -113,7 +116,18 @@ public File getConfigDir() {
      */
     @Incubating
     public void setConfigDir(File configDir) {
-        this.configDir = configDir;
+        this.configDir.set(configDir);
```
Comment (big-guy) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java]: I think we should `@Deprecated` these methods.

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java [hunk 13]
```diff
@@ -113,7 +116,18 @@ public File getConfigDir() {
...
+    }
+
+    /**
+     * Gets the configuration directory.
+     *
+     * @return The configuration directory
+     * @since 4.7
+     */
+    @Incubating
+    public DirectoryProperty getConfigDirectory() {
```
Comment (big-guy) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java]: We need to add this to `org.gradle.api.plugins.quality.CheckstyleExtension.xml`

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java [hunk 14]
```diff
@@ -51,8 +55,7 @@ protected String getToolName() {
     protected CodeQualityExtension createExtension() {
         extension = project.getExtensions().create("checkstyle", CheckstyleExtension.class, project);
         extension.setToolVersion(DEFAULT_CHECKSTYLE_VERSION);
-
-        extension.setConfigDir(project.file("config/checkstyle"));
+        extension.getConfigDirectory().set(determineConfigurationDirectory());
```
Comment (big-guy) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java]: We should fix the call to the Checkstyle task below to use this property.

```
task.setConfigDir(project.provider(new Callable<File>() {
            @Override
            public File call() {
                return extension.getConfigDir();
            }
        }));
```
becomes
```
task.setConfigDir(extension.getConfigDirectory());
```

We should clean-up the Checkstyle task too (to follow our Provider API recommendations).

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java (hunk 13 above)
Comment (bmuschko) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java]: done

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java (hunk 12 above)
Comment (bmuschko) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstyleExtension.java]: I think the idea was to do that a little later: https://github.com/gradle/gradle/pull/4507#discussion_r173125501

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java (hunk 14 above)
Comment (bmuschko) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/CheckstylePlugin.java]: I think this could easily be done in a separate PR.
